**Note: Remember to copy this exam into your Google drive before beginning!**

# Fall 2023 Midterm

Welcome to your Fall 2023 Midterm Exam! Throughout the entire exam, your questions will be *thematic* such that each question should feel like a natural follow-up that a business executive or product manager would ask of you about your work. If you can work through this exam sufficiently, then you are well on your way to entering the workforce!

Your exam questions will be both qualitative *and* quantitative—just as your work in real-life will be! As you work through your exam, be sure to *carefully* read through every question before answering it.

This is a closed-book exam. The only time your screen should be off the `.ipynb` is if it's in canvas, or in the documentation I have provided to you below. I will be walking around to check, fam.

**You have 1 hour and 15 minutes** to take your test and submit in Canvas in the `Midterm` link in the `Assignments` tab. Be sure that you **give yourself at least 5 minutes at the end of class** to ensure you have time to submit your exam.

Deep breaths. You've got this.

General documentation you are allowed to use:

* [ISLP Documentation](https://islp.readthedocs.io/en/latest/installation.html)
* [Pandas Documentation](https://pandas.pydata.org/docs/reference/index.html)
*   [Altair Documentation](https://altair-viz.github.io/user_guide/marks/index.html)
* [scikit-learn Documentation](https://scikit-learn.org/stable/user_guide.html#)
* [statsmodels Documentation](https://www.statsmodels.org/stable/user-guide.html)

# Exam Start

This is it. You just graduated Chapman University `super-mega magna cum laude`, were given the key to the city by the Mayor of Orange, and were hired as a business analyst at the most exclusive, major finance firm ever: Atlas Worldwide Organized Outlay Finance (AWOOF). So life is going pretty great.

It's your first day at work. You've arrived at AWOOF headquarters, walked up to your desk, and laying there is a piece of paper with some instructions for you:

```
# Welcome to AWOOF, Whiz Kid! 🌍💼✨

Dear [Your Name Here],

First off, a super-sized, humongous, jumbo-tron CONGRATULATIONS! 🎉🎉  
Being a *super-mega magna cum laude* graduate is no small feat,
nor is getting the key to a whole city! (Do remember to lock up after you're done.)

I'm Mr. Bazzillion Bucks, the CEO of **AWOOF**. I've personally
waited 17 years, 3 months, 27 days, 9 hours, 32 minutes, and... oh, about 17 seconds, to welcome someone
of your caliber. So let's not waste any more precious milliseconds, shall we?

I need you to get to work making money so I get buy more dog toys... I mean diamonds.

Get to work.

Doggedly Yours,

🖋️  
**Mr. Bazzillion Bucks**  
CEO of Atlas Worldwide Organized Outlay Finance
```

Below the letter are a list of tasks from Mr. Bazzillion Bucks. Better get to work!

# Question 1

I want to understand how well we've been investing our money.

a. Install the ISLP package, restart your runtime, load the `Smarket` dataset, and assign the data to the variable `market` (1pts)
  - more info about the dataset found [here](https://islp.readthedocs.io/en/latest/datasets/Smarket.html)

b. I want to understand which years trading `Direction` was more `Up` than `Down`. Group the data by `Year` and get the *counts of the values* by `Direction`. Then simply write in a comment which years were more up than down. (3pts) Hint: after grouping by a column, you can use double brackets `[[ ... ]]`

c. I (Mr. Bazzillion Bucks) want pretty pictures. I don't understand data easily. Create an Altair plot of the data you created in problem b. (3pts)
  - You should first *group* the original `market` data *by* Year, and get the value counts of the `Direction` data. You may need to reset the columns and rename the column. Assign that new DataFrame to a `yearly_market` variable. Hint: you can immediately rename a column like so: `reset_index(name="some_name")`
  - On the x axis plot the year, on the y axis plot the name of your newly created column. **Make the chart a bar chart.**
    - Ensure that `Year` is the correct data type before plotting! If your chart is wrong below, it will be marked down. See: [DataFrame.astype()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html)

  - assign the 'Direction` column as the color in the encode step so I can see which data is down and up.


You are allowed to use the [altair documentation](https://altair-viz.github.io/user_guide/marks/index.html). But only the altair documentation.


In [52]:
# a
! pip install ISLP

In [10]:
# a
from ISLP import load_data

market = load_data("Smarket")

In [11]:
# b
# Which years were up more than down?

result = (market.groupby("Year")[["Direction"]].value_counts().reset_index(name="Count"))
result

# 2003, 2004, 2005

Year Direction  Count
0  2001      Down    125
1  2001        Up    117
2  2002      Down    140
3  2002        Up    112
4  2003        Up    138
5  2003      Down    114
6  2004        Up    140
7  2004      Down    112
8  2005        Up    141
9  2005      Down    111

In [13]:
# c
import altair as alt

yearly_market = market.groupby("Year")[['Direction']].value_counts().reset_index(name="Count")
yearly_market['Year'] = yearly_market['Year'].astype(str)

(
    alt.Chart(yearly_market)
       .mark_bar()
       .encode(x='Year', y='Count', color='Direction')
)

alt.Chart(...)

*After emailing Mr. Bazzillion Bucks your results, he only has more questions! Here is what he says:*


# Question 2

That chart was confusing! It was too small. And also that data didn't represent today. Let's try something else.

- a. Load the `Fund` data from the ISLP package [here](https://islp.readthedocs.io/en/latest/datasets/Fund.html). Assign it to the variable `fund`. This is information about the managers I pay to trade my money around. I suspect that some are making systemically great trades. I want you to identify them. (1pt)

- b. Take the median of every fund manager. *sort* the *values*. Assign this result to the variable `median_vals`. run `describe` on `median_vals`. Assign it to the variable `dist` and display its results. (2pts)


- c. The `Empirical Rule`, or the `68-95-99.7 Rule`, says that around 68% of results should lie within 1 standard deviation of the mean, around 95% of scores within 2 standard deviations of the mean, and around 99.7% of scores should lie within 3 standard deviations of the mean, if we assume a normal distribution.

  Assign the `mean` from `dist` to variable `mu`, and assign standard deviation from `dist` to the variable `sig`. Calculate what three standard deviations above the mean should be, assign it to the variable `three_sig_above`. (3pts)

- d. Filter `median_vals` down to only include data that is greater than `three_sig_above`. Display who the underperformers are by printing out the filtered `median_vals` results. These are the managers who are systemically overperforming... They will be rewarded. (2pts)

In [43]:
# a
fund = load_data("Fund")

# b
median_vals = fund.median().sort_values()

dist = median_vals.describe()
dist

count    2000.000000
mean       -0.015550
std         1.549189
min        -5.961558
25%        -0.899423
50%        -0.038731
75%         0.830911
max         6.227548
dtype: float64

In [44]:
# c
mu = dist['mean']
sig = dist['std']
three_sig_above= mu + (3*sig)

three_sig_above

4.632015963235064

In [45]:
# d: systemic overperformers
median_vals[median_vals > three_sig_above]

Manager118     4.799317
Manager1602    4.948836
Manager57      4.999272
Manager192     5.018994
Manager91      5.248800
Manager106     5.531031
Manager68      5.684512
Manager75      5.735545
Manager180     5.970380
Manager195     6.227548
dtype: float64

*Well, you've sent your results to Mr. Bazzillion Bucks and he once again has more work for you...*

# Question 3

You did so well at identifying great fund managers, I thought maybe you could help me with something more advanced... One of the companies I'm thinking of buying has wage and demographic data of its workers found [here](https://islp.readthedocs.io/en/latest/datasets/Wage.html). I want to understand the trend of wages over time.

- a. Load the `Wage` data into a variable `wages`. Sort the data by `Year`. Assign the sorted DataFrame back to `wages`. (1pt)

- b. The data is pretty messy. We're going to have to fix that. (6pts)
  - create a new `ever_college` column out of all text data in the column `education`. Assign a value to 1 if someone had any college or advanced education at all and 0 someone had only high school or less. (2pts)
  - create a new `good_health` column out of all text data in the column `health`. Assign a value of 1 if someone has very good health or better, and a value of 0 if someone only had good health or less. (2pts)
  - create a new `is_old` column out of all the numeric data in the column `age`. Assign a value of 1 if someone is 50 years of age or older, assign a value of 0 if someone is less than 50 (2pts).

- c. [Split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) the data into training and test datasets. The training dataset should be 82% of the original data and the test dataset should be 18% of the original data. Assign them into variables `X_train, X_test, y_train, y_test`. The indepdenent variables should be: `year`, `age`, `ever_college`, `good_health`, `is_old`. The dependent variable should be: `logwage`. (3pts)
  - *careful how you split your data here* this data is time sensitive so it cannot be treated randomly...

- d. Please run a linear regression on the training data using the [`statsmodels.api` ols module](https://www.statsmodels.org/stable/regression.html). Make sure there is an intercept term/constant. Print out the summary results. (2pts)

- e. Stats and Coefficients (8pts):
  - Which independent variables are statistically significant at the five percent level? (2pts)
  - Explain to me the expected relationship between those who are classified as is_old and logwage using the coefficient values provided and the language of a statistician. Hint: since `is_old` is a binary variable, you take its exponential before interpreting it. (3pts)
  - If the goal was to predict logwages of workers, would you give this model to Mr Bazzillion Bucks? Why or why not? (Justify your answer rigorously). (3pts)

In [46]:
wages = load_data("Wage").sort_values('year').reset_index(drop=True)
wages.head()

year  age       maritl      race        education              region  \
0  2003   39   2. Married  1. White       2. HS Grad  2. Middle Atlantic   
1  2003   58   2. Married  1. White     1. < HS Grad  2. Middle Atlantic   
2  2003   61  4. Divorced  1. White       2. HS Grad  2. Middle Atlantic   
3  2003   29   2. Married  3. Asian       2. HS Grad  2. Middle Atlantic   
4  2003   25   2. Married  1. White  3. Some College  2. Middle Atlantic   

        jobclass          health health_ins   logwage        wage  
0  1. Industrial       1. <=Good      2. No  4.144605   63.092717  
1  1. Industrial       1. <=Good      2. No  4.579784   97.493294  
2  1. Industrial  2. >=Very Good     1. Yes  5.000000  148.413159  
3  1. Industrial  2. >=Very Good     1. Yes  4.556303   95.230713  
4  1. Industrial  2. >=Very Good     1. Yes  4.477121   87.981033

In [47]:
wages['education'].value_counts()

2. HS Grad            971
4. College Grad       685
3. Some College       650
5. Advanced Degree    426
1. < HS Grad          268
Name: education, dtype: int64

In [48]:
# ever_college

# different ways of achieving the same thing:
ever_college = [0 if i in ('2. HS Grad', '1. < HS Grad') else 1 for i in wages['education']]

# mapping = {'2. HS Grad': 0, '1. < HS Grad': 0, 'College Grad': 1, '3. Some College': 1, '5. Advanced Degree': 1}
# ever_college = wages['education'].map(mapping).tolist()

wages['ever_college'] = ever_college


In [49]:
wages['health'].value_counts()

2. >=Very Good    2142
1. <=Good          858
Name: health, dtype: int64

In [50]:
# good_health

# different ways of achieving the same thing:
good_health = [0 if i in ('1. <=Good') else 1 for i in wages['health']]

# mapping = {'1. <=Good': 0, '2. >=Very Good': 1}
# good_health = wages['health'].map(mapping).tolist()
# or
# good_health = wages['health'].apply(lambda x: 0 if x == '1. <=Good' else 1).tolist()

wages['good_health'] = good_health

In [52]:
# is_old

# different ways of achieving the same thing:
is_old = [0 if i < 50 else 1 for i in wages['age']]

# is_old = wages['age'].apply(lambda x: 0 if x < 50 else 1).tolist()

wages['is_old'] = is_old

In [53]:
from IPython.utils.path import random
from sklearn.model_selection import train_test_split

X = wages[['year','age', 'ever_college', 'good_health','is_old']]
y = wages['logwage']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.18, shuffle=False, random_state=5892)

In [54]:
import statsmodels.api as sm

X_train = sm.add_constant(X_train)

reg = sm.OLS(y_train, X_train).fit()

print(reg.summary())


                            OLS Regression Results                            
Dep. Variable:                logwage   R-squared:                       0.233
Model:                            OLS   Adj. R-squared:                  0.232
Method:                 Least Squares   F-statistic:                     149.5
Date:                Sun, 12 Nov 2023   Prob (F-statistic):          7.34e-139
Time:                        22:20:52   Log-Likelihood:                -584.69
No. Observations:                2460   AIC:                             1181.
Df Residuals:                    2454   BIC:                             1216.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const          -29.8416      7.615     -3.919   

 e. Stats and cofficients
 1. Every variable, including the intercept is statistically sigificant at least at the 5% level.
 2. Compared to those who are not classified as "old" in our model, for those who are 50 years of age or older our model expects that they are associated with 88.3% of log wages compared to those younger than 50 or a 100*(1 - 0.883) = 12.11% decrease in the expected log wage of 50+ year old works, holding all else constant.
 3. No. The R2 is incredibly low at .233 meaning the model does not capture much variance in the training data. It's hard to imagine it would capture **more** variance out of sample. We can confidently claim this model is underfit.


In [56]:
from math import exp
exp(-0.12391)

0.883459348646226

*After you've emailed your results to Mr. Bazzillion Bucks, he responded with a final set of questions for you*

# Question 4

Great! You've answered all my questions. Answer these last questions for me and I will exit my office and reveal myself to you!

We are thinking of getting into the credit game. Assess this credit dataset for me and tell me what you think.

- a. Load the [`Credit`](https://islp.readthedocs.io/en/latest/datasets/Credit.html) dataset from ISLP. Assign it to the `credit` variable. Create a target variable for us: whenever the column `Limit` is above 6000 assign a 1, otherwise assign a 0. Assign these results to a new column: `is_high`. (2pts)
- b. Create new features for us. In the `Gender` column map those who are assigned ' Male' to 1 and those who are assigned 'Female' 0. Assign the new data to the column `is_male` (2pts). Hint: The Male column has a typo...
- c. Create another new feature column. In the `Education` column assign values 16 and above to 1, and all other values to 0. Assign the results to the new column `is_educated`. (2pts)
- d. Run a train_test_split on the credit dataset. The feature data should be: `'Income', 'Rating', 'Cards', 'Age', 'Education', 'Balance', 'is_male', 'is_educated' `. The target should be `is_high` from above. The test_size should be .17. The data should be shuffled and a `random_state` integer should be given so I can recreate your work. (2pts)
- e. Run a [LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) on your `X_train` and `y_train` data. Then predict on your `X_test` data. (2pts)
- f. Run a [confusion matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html) on your results. Interpret your results for me. Interpret every cell in the matrix. (4pts)
- g. Run an [ROC/AUC score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score) on your data. Are the results good? How do you know? Would you put this model into production in the real world? Why or why not? Intepret the AUC value for us. (4pts)

In [66]:
# a

credit = load_data('Credit')
credit['is_high'] = (credit['Limit'] > 6000).astype(int)

# # another method
# # credit['is_high'] = credit['Rating'].map(lambda x: 1 if x > 6000 else 0)

# credit.shape

In [82]:
# b
gender_mapping = {" Male": 1, "Female": 0}
credit['is_male'] = credit['Gender'].map(gender_mapping)

In [84]:
# c

credit['is_educated'] = [1 if i >= 16 else 0 for i in credit['Education']]

In [85]:
# d.
X = credit[['Income', 'Rating', 'Cards', 'Age', 'Education', 'Balance', 'is_male', 'is_educated']]
y = credit['is_high']


In [87]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.17,
                                                    random_state=643, shuffle=True)

In [88]:
#e.
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [89]:
y_pred = log_reg.predict(X_test)

In [90]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[45,  0],
       [ 4, 19]])

- The results are pretty great! We have 19 True Negatives and 45 True Positives. predicted our True Positives and True Negatives in our out of sample prediction. That means the model generalizes very well. We predicted 4 False Negatives and 0 False Positive.

In [91]:
from sklearn.metrics import roc_auc_score

roc_auc_score(y_test, y_pred)

0.9130434782608696

In [ ]:
# g

# The results are high at ~.91. The best we can hope for is 1.0, the worst is 0.
# I still wonder if we should put it into production because in the real world,
# data is going to be messier and we only trained on 400 data points.
# There's always going to be exceptions we haven't seen, but 400 seems unreasonably too low
# It can lead one to wonder if the model is overfit. We probably want to test on more data.
# Still, if someone did put it into production, it wouldn't be the worst in the world.

As you finish your results and send them in, Mr. Bazzillion Bucks exits his CEO office and you get to see who he is for the first time... Run the code below to find out who this myster CEO is!

In [ ]:
# Run this code if you have time, otherwise submit your notebook!
from IPython.display import Image, display

print("Mr Bazzillion Bucks was Atlas!")

image_url = "https://user-images.githubusercontent.com/45580612/275377290-33ebb2ba-37bd-4d20-a9cd-f8ae70fb21ed.jpeg"
display(Image(url=image_url, width=450, height=500))

Mr Bazzillion Bucks was Atlas!


### Submit your exam in Canvas!